include/xtypes/DynamicData.hpp

```diff
@@ -235,6 +235,173 @@ class ReadableDynamicDataRef
         catch(bool value) { return value; }
     }
 
+    /// \brief Class used for iterate ReadableDynamicDataRef
+    class Iterator
+    {
+    public:
+        Iterator(
+                const ReadableDynamicDataRef& ref)
+            : Iterator(ref, false)
+        {}
+
+        Iterator& operator = (
+                const Iterator& other)
+        {
+            data_ = other.data_;
+            index_ = other.index_;
+            return *this;
+        }
+
+        bool operator == (
+                const Iterator& other) const
+        {
+            return other.data_ == data_ && other.index_ == index_;
+        }
+
+        bool operator != (
+                const Iterator& other) const
+        {
+            return other.data_ != data_ || other.index_ != index_;
+        }
+
+        ReadableDynamicDataRef operator * () const
+        {
+            const CollectionType& collection = static_cast<const CollectionType&>(type_);
+            return ReadableDynamicDataRef(collection.content_type(), collection.get_instance_at(data_, index_));
+        }
+
+        Iterator& operator ++ ()
+        {
+            ++index_;
+            return *this;
+        }
+
+        Iterator operator ++ (
+                int)
+        {
+            Iterator prev = *this;
+            ++index_;
+            return prev;
+        }
+
+    protected:
+
+        friend class ReadableDynamicDataRef;
+
+        Iterator(
```

Comment:
Is there some way to avoid that the user could call this constructor? Only must be accessible by the begin/end functions. (The user could create an Iterator` from any `DynamicData` without control).